### PR TITLE
Add header for openPMD-specific errors

### PR DESCRIPTION
**Pull request by franzpoeschel**
This is just the scaffolding for the more fine-grained error-handling that we want to implement.
This adds:
* A class `openPMD::Error` which should serve as base class for all of openPMD's error types. It should not be instatiated directly, but via child classes.
* A namespace `openPMD::error` which bundles all those error types.
* An example error type `openPMD::error::OperationUnsupportedInBackend`, without using it.

Related: #983 (this is not a fix, but the fundament for a fix)

TODO:
- [x] Python?

### CODE REVIEW

File: include/openPMD/Error.hpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+private:
+    std::string m_what;
+
+protected:
+    Error( std::string what ) : m_what( what )
+    {
+    }
+
+public:
+    virtual const char * what() const noexcept final;
```
Comment (ax3l) [inline include/openPMD/Error.hpp]: Intentionally `final` though?

File: include/openPMD/Error.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * in the openPMD::error namespace.
+ *
+ */
+class Error : public std::exception
+{
+private:
+    std::string m_what;
+
+protected:
+    Error( std::string what ) : m_what( what )
```
Comment (ax3l) [inline include/openPMD/Error.hpp:22]: We might want to default the other constructors as well here:
https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)#Rule_of_Five

File: src/binding/python/Error.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    auto & baseError = py::register_exception< Error >( m, "Error" );
+    py::register_exception< error::OperationUnsupportedInBackend >(
+        m, "ErrorOperationUnsupportedInBackend", baseError );
+
+#ifndef NDEBUG
+    m.def( "test_throw", []( std::string description ) {
+        throw error::OperationUnsupportedInBackend( "json", description );
+    } );
+#endif
+}
```
Comment (ax3l) [inline src/binding/python/Error.cpp]:
```suggestion
}

```

File: include/openPMD/Error.hpp (hunk 1 above)
Comment (franzpoeschel) [reply include/openPMD/Error.hpp]: I probably had a reason for it, but I can't remember it and I can't come up with one any more, so I'll take it out again.

**Review by ax3l — APPROVED** [review]
Thanks a lot! :tada:

**Pull request merged**